### ISSUE TITLE
OS templates "Trisquel 10", and "Trisquel 11": Cannot find USB devices

### ISSUE THREAD

**Issue by davidhedlund** [virt-manager/virt-manager#589]
Trisquel 10 is based on Ubuntu 20.04, and Trisquel 11 is based on Ubuntu 22.04.

| ISO |  OS template during installation         | DISK and CDROM (selected from the OS template)    | Disk mount point |  lsusb after "Redirect USB device"[1]  | USB mount point
|---: | ---: | :---------------------------- |:-------------:| -----:| --:|
| trisquel_10.0_amd64.iso | Trisquel 10 LTS           | IDE (hda), IDE CDROM (hdb) | /dev/sda[1-4] | Not found | - |
| trisquel_10.0_amd64.iso | Ubuntu 20.04 LTS           | VirtIO (vda), SATA CDROM (sda) | /dev/vda[1-4] | Found | /dev/sda |
| trisquel_11.0_amd64.iso | Trisquel 11 LTS           | IDE (hda), IDE CDROM (hdb) | /dev/sda[1-4] | Not found | - |
| trisquel_11.0_amd64.iso | Ubuntu 22.04 LTS           | VirtIO (vda), SATA CDROM (sda) | /dev/vda[1-4] | Found | /dev/sda |
| ubuntu-20.04.6-desktop-amd64.iso | Ubuntu 20.04 LTS           | VirtIO (vda), SATA CDROM (sda) | /dev/vda[1-4] | Found | /dev/sda |
| ubuntu-22.04-desktop-amd64.iso | Ubuntu 22.04 LTS           | VirtIO (vda), SATA CDROM (sda) | /dev/vda[1-4] | Found | /dev/sda |

1: virt-manager: From either "Virtual Machine -> Redirect USB device", or by adding "Show virtual hardware Details -> Hardware -> USB Host Device". Then `lsusb` in Ubuntu confirmed that the device is detected.

# Suggested solution
Based on the findings in the table above:
* The OS template "Ubuntu 20.04 LTS" could be copied to "Trisquel 10 LTS"
* The OS template "Ubuntu 22.04 LTS" could be copied to "Trisquel 11 LTS"

# virt-manager package info
```
$ apt-cache show virt-manager
Package: virt-manager
Architecture: all
Version: 1:4.0.0-1
Priority: optional
Section: universe/admin
Origin: Ubuntu
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Debian Libvirt Maintainers <pkg-libvirt-maintainers@lists.alioth.debian.org>
Bugs: https://bugs.launchpad.net/ubuntu/+filebug
Installed-Size: 2298
Depends: gir1.2-gtk-3.0 (>= 3.10), gir1.2-gtk-vnc-2.0, gir1.2-gtksource-4, gir1.2-libosinfo-1.0, gir1.2-libvirt-glib-1.0, gir1.2-vte-2.91, python3-dbus, python3-gi (>= 3.31.3~), python3-gi-cairo, python3-libvirt (>= 0.7.1), virtinst (= 1:4.0.0-1), dconf-gsettings-backend | gsettings-backend, python3:any
Recommends: gir1.2-ayatanaappindicator3-0.1 | gir1.2-appindicator3-0.1, gir1.2-spiceclientglib-2.0, gir1.2-spiceclientgtk-3.0, libvirt-daemon-system (>= 1.2.7)
Suggests: gir1.2-secret-1, gnome-keyring, python3-guestfs, ssh-askpass, virt-viewer
Filename: pool/main/v/virt-manager/virt-manager_4.0.0-1_all.deb
SHA512: b6618d34087f2f5edb7fb19ce5b7e7fa151c4190b82ce5360b03050229837996d589af7965f5509c6a6570a605d17780733a57e9f4e76880cce8d4cf627fb0aa
Homepage: https://virt-manager.org/
Size: 261800
SHA256: 5f54b22d8dd4dd2db7e82f5267e72981699d97600a9ee56bfd98a6f2ea58538c
SHA1: 72b92481fcccaffaf2a947da85ae6f772ec8b258
MD5sum: 548c6d54a2aab650eb87db324a8ba502
```

**Comment by ptoscano** [comment]
The definitions of the devices to use in VMs mostly come from the `osinfo-db` package, which contains the database of the libosinfo library that virt-manager uses.

> Based on the findings in the table above:
> * The OS template "Ubuntu 20.04 LTS" could be copied to "Trisquel 10 LTS"
> * The OS template "Ubuntu 22.04 LTS" could be copied to "Trisquel 11 LTS"

This is already the case in `osinfo-db`: `http://trisquel.info/trisquel/10` (the ID of the OS representing Trisquel 10) `derives-from` `http://ubuntu.com/ubuntu/20.04`, and indeed the devices are properly inherited:
```
$ OSINFO_SYSTEM_DIR=$PWD/data ../list-devices.py http://ubuntu.com/ubuntu/20.04 2>/dev/null 
 name              | class          | subsystem   | vendor                   | product                                         | bus
-------------------+----------------+-------------+--------------------------+-------------------------------------------------+-------
 e1000e            | net            | -           | [8086] Intel Corporation | [10d3] 82574L Gigabit Network Connection        | pci
 ich9-hda          | audio          | -           | [8086] Intel Corporation | [293e] 82801I (ICH9 Family) HD Audio Controller | pci
 nec-xhci          | controller.usb | -           | [1033] NEC Corporation   | [0194] uPD720200 USB 3.0 Host Controller        | pci
 qemu-x86-q35      | chipset        | -           | -                        | -                                               | -
 qemu-xhci         | controller.usb | -           | [1b36] Red Hat, Inc.     | [0004] QEMU PCI Quad-port 16550A Adapter        | pci
 tablet            | input          | -           | [80ee] VirtualBox        | [0021] USB Tablet                               | usb
 virtio-block      | block          | -           | [1af4] Red Hat, Inc.     | [1001] Virtio block device                      | pci
 virtio-net        | net            | -           | [1af4] Red Hat, Inc.     | [1000] Virtio network device                    | pci
 virtio1.0-9p      | filesystem     | -           | [1af4] Red Hat, Inc.     | [1049] Virtio 1.0 filesystem                    | pci
 virtio1.0-balloon | memory.balloon | -           | [1af4] Red Hat, Inc.     | [1045] Virtio 1.0 memory balloon                | pci
 virtio1.0-block   | block          | -           | [1af4] Red Hat, Inc.     | [1042] Virtio 1.0 block device                  | pci
 virtio1.0-console | console        | -           | [1af4] Red Hat, Inc.     | [1043] Virtio 1.0 console                       | pci
 virtio1.0-gpu     | video          | -           | [1af4] Red Hat, Inc.     | [1050] Virtio 1.0 GPU                           | pci
 virtio1.0-input   | input          | -           | [1af4] Red Hat, Inc.     | [1052] Virtio 1.0 input                         | pci
 virtio1.0-net     | net            | -           | [1af4] Red Hat, Inc.     | [1041] Virtio 1.0 network device                | pci
 virtio1.0-rng     | rng            | -           | [1af4] Red Hat, Inc.     | [1044] Virtio 1.0 RNG                           | pci
 virtio1.0-scsi    | block          | -           | [1af4] Red Hat, Inc.     | [1048] Virtio 1.0 SCSI                          | pci
$ OSINFO_SYSTEM_DIR=$PWD/data ../list-devices.py http://trisquel.info/trisquel/10 2>/dev/null 
 name              | class          | subsystem   | vendor                   | product                                         | bus
-------------------+----------------+-------------+--------------------------+-------------------------------------------------+-------
 e1000e            | net            | -           | [8086] Intel Corporation | [10d3] 82574L Gigabit Network Connection        | pci
 ich9-hda          | audio          | -           | [8086] Intel Corporation | [293e] 82801I (ICH9 Family) HD Audio Controller | pci
 nec-xhci          | controller.usb | -           | [1033] NEC Corporation   | [0194] uPD720200 USB 3.0 Host Controller        | pci
 qemu-x86-q35      | chipset        | -           | -                        | -                                               | -
 qemu-xhci         | controller.usb | -           | [1b36] Red Hat, Inc.     | [0004] QEMU PCI Quad-port 16550A Adapter        | pci
 tablet            | input          | -           | [80ee] VirtualBox        | [0021] USB Tablet                               | usb
 virtio-block      | block          | -           | [1af4] Red Hat, Inc.     | [1001] Virtio block device                      | pci
 virtio-net        | net            | -           | [1af4] Red Hat, Inc.     | [1000] Virtio network device                    | pci
 virtio1.0-9p      | filesystem     | -           | [1af4] Red Hat, Inc.     | [1049] Virtio 1.0 filesystem                    | pci
 virtio1.0-balloon | memory.balloon | -           | [1af4] Red Hat, Inc.     | [1045] Virtio 1.0 memory balloon                | pci
 virtio1.0-block   | block          | -           | [1af4] Red Hat, Inc.     | [1042] Virtio 1.0 block device                  | pci
 virtio1.0-console | console        | -           | [1af4] Red Hat, Inc.     | [1043] Virtio 1.0 console                       | pci
 virtio1.0-gpu     | video          | -           | [1af4] Red Hat, Inc.     | [1050] Virtio 1.0 GPU                           | pci
 virtio1.0-input   | input          | -           | [1af4] Red Hat, Inc.     | [1052] Virtio 1.0 input                         | pci
 virtio1.0-net     | net            | -           | [1af4] Red Hat, Inc.     | [1041] Virtio 1.0 network device                | pci
 virtio1.0-rng     | rng            | -           | [1af4] Red Hat, Inc.     | [1044] Virtio 1.0 RNG                           | pci
 virtio1.0-scsi    | block          | -           | [1af4] Red Hat, Inc.     | [1048] Virtio 1.0 SCSI                          | pci
```
(`list-devices.py` is a small script of mine that queries the devices available for an OS using libosinfo.)

As you can see, the two OS have the same available devices. The same applies to Ubuntu 22.04 & Trisquel 11.

Note that Trisquel 10 & 11 are available in `osinfo-db` starting from version 20230719.

What's the version of the `osinfo-db` package installed?

**Comment by davidhedlund** [comment]
> The definitions of the devices to use in VMs mostly come from the `osinfo-db` package, which contains the database of the libosinfo library that virt-manager uses.
> 
> > Based on the findings in the table above:
> > 
> > * The OS template "Ubuntu 20.04 LTS" could be copied to "Trisquel 10 LTS"
> > * The OS template "Ubuntu 22.04 LTS" could be copied to "Trisquel 11 LTS"
> 
> This is already the case in `osinfo-db`: `http://trisquel.info/trisquel/10` (the ID of the OS representing Trisquel 10) `derives-from` `http://ubuntu.com/ubuntu/20.04`, and indeed the devices are properly inherited:
> 
> ```
> $ OSINFO_SYSTEM_DIR=$PWD/data ../list-devices.py http://ubuntu.com/ubuntu/20.04 2>/dev/null 
>  name              | class          | subsystem   | vendor                   | product                                         | bus
> -------------------+----------------+-------------+--------------------------+-------------------------------------------------+-------
>  e1000e            | net            | -           | [8086] Intel Corporation | [10d3] 82574L Gigabit Network Connection        | pci
>  ich9-hda          | audio          | -           | [8086] Intel Corporation | [293e] 82801I (ICH9 Family) HD Audio Controller | pci
>  nec-xhci          | controller.usb | -           | [1033] NEC Corporation   | [0194] uPD720200 USB 3.0 Host Controller        | pci
>  qemu-x86-q35      | chipset        | -           | -                        | -                                               | -
>  qemu-xhci         | controller.usb | -           | [1b36] Red Hat, Inc.     | [0004] QEMU PCI Quad-port 16550A Adapter        | pci
>  tablet            | input          | -           | [80ee] VirtualBox        | [0021] USB Tablet                               | usb
>  virtio-block      | block          | -           | [1af4] Red Hat, Inc.     | [1001] Virtio block device                      | pci
>  virtio-net        | net            | -           | [1af4] Red Hat, Inc.     | [1000] Virtio network device                    | pci
>  virtio1.0-9p      | filesystem     | -           | [1af4] Red Hat, Inc.     | [1049] Virtio 1.0 filesystem                    | pci
>  virtio1.0-balloon | memory.balloon | -           | [1af4] Red Hat, Inc.     | [1045] Virtio 1.0 memory balloon                | pci
>  virtio1.0-block   | block          | -           | [1af4] Red Hat, Inc.     | [1042] Virtio 1.0 block device                  | pci
>  virtio1.0-console | console        | -           | [1af4] Red Hat, Inc.     | [1043] Virtio 1.0 console                       | pci
>  virtio1.0-gpu     | video          | -           | [1af4] Red Hat, Inc.     | [1050] Virtio 1.0 GPU                           | pci
>  virtio1.0-input   | input          | -           | [1af4] Red Hat, Inc.     | [1052] Virtio 1.0 input                         | pci
>  virtio1.0-net     | net            | -           | [1af4] Red Hat, Inc.     | [1041] Virtio 1.0 network device                | pci
>  virtio1.0-rng     | rng            | -           | [1af4] Red Hat, Inc.     | [1044] Virtio 1.0 RNG                           | pci
>  virtio1.0-scsi    | block          | -           | [1af4] Red Hat, Inc.     | [1048] Virtio 1.0 SCSI                          | pci
> $ OSINFO_SYSTEM_DIR=$PWD/data ../list-devices.py http://trisquel.info/trisquel/10 2>/dev/null 
>  name              | class          | subsystem   | vendor                   | product                                         | bus
> -------------------+----------------+-------------+--------------------------+-------------------------------------------------+-------
>  e1000e            | net            | -           | [8086] Intel Corporation | [10d3] 82574L Gigabit Network Connection        | pci
>  ich9-hda          | audio          | -           | [8086] Intel Corporation | [293e] 82801I (ICH9 Family) HD Audio Controller | pci
>  nec-xhci          | controller.usb | -           | [1033] NEC Corporation   | [0194] uPD720200 USB 3.0 Host Controller        | pci
>  qemu-x86-q35      | chipset        | -           | -                        | -                                               | -
>  qemu-xhci         | controller.usb | -           | [1b36] Red Hat, Inc.     | [0004] QEMU PCI Quad-port 16550A Adapter        | pci
>  tablet            | input          | -           | [80ee] VirtualBox        | [0021] USB Tablet                               | usb
>  virtio-block      | block          | -           | [1af4] Red Hat, Inc.     | [1001] Virtio block device                      | pci
>  virtio-net        | net            | -           | [1af4] Red Hat, Inc.     | [1000] Virtio network device                    | pci
>  virtio1.0-9p      | filesystem     | -           | [1af4] Red Hat, Inc.     | [1049] Virtio 1.0 filesystem                    | pci
>  virtio1.0-balloon | memory.balloon | -           | [1af4] Red Hat, Inc.     | [1045] Virtio 1.0 memory balloon                | pci
>  virtio1.0-block   | block          | -           | [1af4] Red Hat, Inc.     | [1042] Virtio 1.0 block device                  | pci
>  virtio1.0-console | console        | -           | [1af4] Red Hat, Inc.     | [1043] Virtio 1.0 console                       | pci
>  virtio1.0-gpu     | video          | -           | [1af4] Red Hat, Inc.     | [1050] Virtio 1.0 GPU                           | pci
>  virtio1.0-input   | input          | -           | [1af4] Red Hat, Inc.     | [1052] Virtio 1.0 input                         | pci
>  virtio1.0-net     | net            | -           | [1af4] Red Hat, Inc.     | [1041] Virtio 1.0 network device                | pci
>  virtio1.0-rng     | rng            | -           | [1af4] Red Hat, Inc.     | [1044] Virtio 1.0 RNG                           | pci
>  virtio1.0-scsi    | block          | -           | [1af4] Red Hat, Inc.     | [1048] Virtio 1.0 SCSI                          | pci
> ```
> 
> (`list-devices.py` is a small script of mine that queries the devices available for an OS using libosinfo.)
> 
> As you can see, the two OS have the same available devices. The same applies to Ubuntu 22.04 & Trisquel 11.

Can you please copy your message (copy/paste, or modify it) to https://github.com/virt-manager/virt-manager/issues/575 ?

> 
> Note that Trisquel 10 & 11 are available in `osinfo-db` starting from version 20230719.
> 
> What's the version of the `osinfo-db` package installed?

Thank you very much for helping me out!

I don't have version 20230719 yet:
```
$ apt show osinfo-db
Package: osinfo-db
Version: 0.20221130-2+11.0trisquel6
Priority: optional
Section: libs
Maintainer: Trisquel GNU/Linux developers <trisquel-devel@listas.trisquel.info>
Installed-Size: 3,585 kB
Homepage: https://libosinfo.org/
Download-Size: 142 kB
APT-Manual-Installed: no
APT-Sources: https://archive.trisquel.org/trisquel aramo-updates/main amd64 Packages
Description: Operating system database files
 libosinfo is a GObject based library API for managing information about
 operating systems, hypervisors and the (virtual) hardware devices they
 can support. It provides APIs to match/identify optimal devices for deploying
 an operating system on a hypervisor.
 .
 This package contains the actual database files.

N: There is 1 additional record. Please use the '-a' switch to see it
```

**Comment by ptoscano** [comment]
> Can you please copy your message (copy/paste, or modify it) to #575 ?

I simply closed that one, because having 2 issues for the same problem makes things confusing, and risks duplicating the discussion (and losing details when doing so).

> Version: 0.20221130-2+11.0trisquel6

So, I took the sources of Debian 0.20221130-2 & Trisquel 0.20221130-2+11.0trisquel6, and compared them. The result shows indeed the problem:

```diff
diff -Nru osinfo-db-0.20221130/data/os/trisquel.info/trisquel-9.xml.in osinfo-db-0.20221130/data/os/trisquel.info/trisquel-9.xml.in
--- osinfo-db-0.20221130/data/os/trisquel.info/trisquel-9.xml.in        2022-11-30 07:19:52.000000000 +0100
+++ osinfo-db-0.20221130/data/os/trisquel.info/trisquel-9.xml.in        2023-06-04 02:27:29.000000000 +0200
@@ -4,13 +4,14 @@
      See http://www.gnu.org/licenses/ for a copy of the license text -->
   <os id="http://trisquel.info/trisquel/9">
     <short-id>trisquel9</short-id>
-    <name>Trisquel</name>
+    <name>Trisquel 9.0 LTS</name>
     <version>9</version>
     <vendor>The Trisquel Project</vendor>
     <family>linux</family>
     <distro>trisquel</distro>
+
     <release-date>2021-10-16</release-date>
-    <derives-from id="http://ubuntu.com/ubuntu/18.04"/>
+    <eol-date>2023-05-30</eol-date>

[...]
diff -Nru osinfo-db-0.20221130/data/os/trisquel.info/trisquel-10.xml.in osinfo-db-0.20221130/data/os/trisquel.info/trisquel-10.xml.in
--- osinfo-db-0.20221130/data/os/trisquel.info/trisquel-10.xml.in       1970-01-01 01:00:00.000000000 +0100
+++ osinfo-db-0.20221130/data/os/trisquel.info/trisquel-10.xml.in       2023-06-04 02:27:29.000000000 +0200
@@ -0,0 +1,84 @@
+<?xml version="1.0"?>
+<libosinfo version="0.0.1">
+  <!-- Licensed under the GNU General Public License version 2 or later.
+     See http://www.gnu.org/licenses/ for a copy of the license text -->
+  <os id="http://trisquel.info/trisquel/10">
+    <short-id>trisquel10</short-id>
+    <short-id>trisquel-nabia</short-id>
+    <name>Trisquel 10.0 LTS</name>
+    <version>10</version>
+    <vendor>The Trisquel Project</vendor>
+    <family>linux</family>
+    <distro>trisquel</distro>
+    <codename>Nabia</codename>
+
+    <release-date>2022-02-01</release-date>
[...]
diff -Nru osinfo-db-0.20221130/data/os/trisquel.info/trisquel-11.xml.in osinfo-db-0.20221130/data/os/trisquel.info/trisquel-11.xml.in
--- osinfo-db-0.20221130/data/os/trisquel.info/trisquel-11.xml.in       1970-01-01 01:00:00.000000000 +0100
+++ osinfo-db-0.20221130/data/os/trisquel.info/trisquel-11.xml.in       2023-06-04 02:27:29.000000000 +0200
@@ -0,0 +1,84 @@
+<?xml version="1.0"?>
+<libosinfo version="0.0.1">
+  <!-- Licensed under the GNU General Public License version 2 or later.
+     See http://www.gnu.org/licenses/ for a copy of the license text -->
+  <os id="http://trisquel.info/trisquel/11">
+    <short-id>trisquel11</short-id>
+    <short-id>trisquel-aramo</short-id>
+    <name>Trisquel 11.0 LTS</name>
+    <version>11</version>
+    <vendor>The Trisquel Project</vendor>
+    <family>linux</family>
+    <distro>trisquel</distro>
+    <codename>Aramo</codename>
+
+    <release-date>2023-03-19</release-date>
[...]
```
In short:
- from Trisquel 9, the existing `<derives-from>` was removed
- in the added/backported Trisquel 10 & 11, `<derives-from>` was omitted (you can compare the actual upstream additions [1][2]

[1] https://gitlab.com/libosinfo/osinfo-db/-/merge_requests/622
[2] https://gitlab.com/libosinfo/osinfo-db/-/merge_requests/623

The removal of the `<derives-from>` relationship means that the OSes do not have *any* of the inheritable details available in the Ubuntu versions they derive from:
- available devices
- firmware information
- potentially also resources (although none of the OS in osinfo-db uses inheritable resources at the moment)

Those changes, from my osinfo-db maintainer POV, are broken, and that's what causes this issue. You saw also in your tests that selecting the Ubuntu OSes instead of the Trisquel ones when installing Trisquel resulted in optimal devices (VirtIO, etc).

While I was checking the diff of the downstream Trisquel changes, I noticed more breakages. I'll show one diffs, although the same issues apply to almost all the OSes:
```diff
     <!-- DVD -->
     <media arch="x86_64">
-      <url>http://archive.fedoraproject.org/pub/archive/fedora/linux/core/2/x86_64/iso/FC2-x86_64-DVD.iso</url>
-      <iso>
         <application-id>FC/2 x86_64</application-id>
         <system-id>LINUX</system-id>
         <volume-size>4273096704</volume-size>
-      </iso>
-      <kernel>isolinux/vmlinuz</kernel>
-      <initrd>isolinux/initrd.img</initrd>
     </media>

     <tree arch="i686">
       <url>http://archive.fedoraproject.org/pub/archive/fedora/linux/core/2/i386/os/</url>
       <kernel>images/pxeboot/vmlinuz</kernel>
       <initrd>images/pxeboot/initrd.img</initrd>
-      <boot-iso>images/boot.iso</boot-iso>
     </tree>
```
In addition to the removal of `<url>` from `<media>`:
- the `<iso>` tags in `<media>` were dropped, "moving" the tags inside it to `<media>`; this breaks the detection of ISO files
- sometimes `<kernel>` and `<initrd>` in `<media>` were dropped too; this breaks the installation of such OSes from their ISOs
- `<boot-iso>` from `<tree>` were dropped; this is only found in Fedora Core 1 to 6 (both included), and it does not seem to be used in virt-manager/virt-install; still, an unneeded downstream change

To add more problems: the two tests that would have pointed out some of the breakages to the OS, i.e. `test_isoinfo.py` and `test_schema.py`, were removed too! This means that, when building a new package and running the test suite, there is literally nothing that can prevent broken changes to be shipped.

To sum up, this is clearly a downstream screwup:
- the available Trisquel OSes are incorrectly severed from the needed derives-from relationship from Ubuntu
  - do not do this, as it will creates an extremely poor experience for Trisquel users
- some of the important tests are removed
  - do not do this, as this works around other screwups
- existing OSes are badly patched
  - I understand the POV of Trisquel to not "want" to show ISOs of other OSes; however, the changes done go way beyond that, and badly break OSes

Closing this bug, as it not a problem in upstream osinfo-db or virt-manager.
